File: Data Structures & Algorithms/same-binary-tree/submission-1.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right

class Solution:
    def isSameTree(self, p: Optional[TreeNode], q: Optional[TreeNode]) -> bool:
        def dfsBothTrees(p,q):
            # Base Cases
            if p == None and q == None:
                return True
            if p == None and q != None or p != None and q == None:
                return False
            
            if p.val == q.val:     
                left_same = dfsBothTrees(p.left, q.left)
                right_same = dfsBothTrees(p.right, q.right)
                return left_same and right_same
                
            else:
                return False

        return dfsBothTrees(p, q)


    

File: Data Structures & Algorithms/binary-tree-diameter/submission-1.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */

class Solution {
public:
    int diameterOfBinaryTree(TreeNode* root) {
        // Solution using recursive DFS (Post Order)
        int diameter = 0;
        dfs(root, diameter);
        return diameter;
        
    }

    int dfs(TreeNode * root, int& diameter) {
        // Function to return height of subtree at 'root'
        if (!root) return 0;

        // Get height of the left and right subtrees
        int leftHeight = dfs(root->left, diameter);
        int rightHeight = dfs(root->right, diameter);

        // Path through current node = leftHeight + rightHeight
        diameter = max(diameter, leftHeight + rightHeight);

        // Return height of this specific subtree to parent
        return 1 + max(leftHeight, rightHeight);


    }
    
};
